### ISSUE TITLE
Best practices for simulation checkpointing and determinism

### ISSUE THREAD

**Issue by JBGreisman** [pandegroup/openmm#2202]
Hi All,

I am planning to run simulations on a cluster that will require multiple allocations to finish. I am hoping to run these simulations on NVIDIA GPUs, and there are different generations of hardware that one can land on with each allocation (though I can always standardize this to land on a single type of GPU, if needed).

I was wondering whether there are any settings for which a simulation loaded from a checkpoint will be equivalent to the same simulation run contiguously. The [documentation](http://docs.openmm.org/latest/userguide/library.html#determinism) mentions certain settings that should yield deterministic simulations using CUDA, but these settings are not yielding deterministic simulations for me when loaded from a checkpoint (i.e. a simulation begun from a checkpoint is not equivalent to the portion of the single, long trajectory after the checkpoint was dropped). I have also tried using the `DeterministicForces` property that I have seen mentioned elsewhere (#2191).

Are there any other current best practices for deterministic simulations on CUDA? To be more specific:

1. Is it possible to get deterministic simulations from a loaded checkpoint that are equivalent to a single, longer trajectory?
2. Are there any scenarios for which the checkpointed simulations can be run on different GPUs with the same CUDA versions?

I'd be happy to put together a minimal example for testing, but I wanted to check if this is a possible use case in the first place.

Thanks,
Jack

**Comment by peastman** [comment]
If you tell it to use deterministic forces then you should generally get identical results.  When you speak of a "checkpoint" are you referring to a file created by calling `Context.createCheckpoint()`, or to something else?

**Comment by JBGreisman** [comment]
Thanks for following up -- I am referring to a file created in that manner, although I am now realizing that I have been using `Simulation.saveCheckpoint()` (not `Context.createCheckpoint()`) to save the file and `Simulation.loadCheckpoint()`(not `Context.loadCheckpoint()`) to load it for restarting simulations.

Is there different behavior between these two sets of methods?

**Comment by peastman** [comment]
No, that's fine.  `Simulation.saveCheckpoint()` is just a convenience wrapper around `Context.createCheckpoint()`.

Can you give more details on exactly what you're doing, and how you're determining that the result is different?  If you can provide code, that would be ideal.  Also, which platform are you using?

For reference, here is an example of a test case that checks for reproducibility after loading a checkpoint: https://github.com/pandegroup/openmm/blob/f9106ddb6a0d72c960c73b23ed0e07cd7190a80f/platforms/cuda/tests/TestCudaCheckpoints.cpp

**Comment by JBGreisman** [comment]
Following up on this thread, if I use a similar testing script to that above and increase the size of the waterbox edges from 25 angstroms to 70 angstroms, I am starting to see differences in position/velocity after 1000 steps. The differences are seen compared to the "reference" simulation, in which the simulation was run contiguously without loading checkpoints. However, the two methods of loading checkpoints that are being tested in the script appear to produce simulations with equivalent positions/velocities. It should be noted though that this does not always happen — occasionally, the 3 simulations all seem to agree in terms of positions/velocities after 10k steps. I am randomly initializing the velocities for the waterbox, so I would expect each running of the test script to produce different simulations, but I am surprised that some appear deterministic and others do not.

I have also uploaded my current testing script along with 2 examples (1 successful after 10k steps and one with position/velocity inconsistencies).

Based on these results it appears that checkpoints are being loaded in a way that is deterministic, but that loading checkpoints is not _always_ producing a simulation with equivalent positions/velocities to if I had run a single, continuous simulation.

For completeness, I am running my simulations using CUDA 9.0 (Driver version: 396.26) on a Tesla K20. I am using the platform properties `DeterministicForces : true` and `Precision : double`, and am currently using OpenMM version 7.2.2 for Python 3.6. 

[testCUDA.zip](https://github.com/pandegroup/openmm/files/2582718/testCUDA.zip)
